ecosystem/ffmpeg_plugin/kahawai_enc.c

```diff
@@ -0,0 +1,396 @@
+  char *dst_addr;
+  int udp_port;
+  int fb_cnt;
+  int session_cnt;
+
+  int width;
+  int height;
+  enum AVPixelFormat pixel_format;
+  AVRational framerate;
+  mtl_handle dev_handle;
+  st20p_tx_handle tx_handle;
+
+  pthread_cond_t get_frame_cond;
+  pthread_mutex_t get_frame_mutex;
+
+  int64_t frame_counter;
+  struct st_frame *frame;
+  size_t output_frame_size;
+} KahawaiMuxerContext;
+
+static mtl_handle shared_st_handle = NULL;
+static unsigned int active_session_cnt = 0;
+static int64_t frame_tx_completed = 0;
+static bool tx_completed = false;
+
+extern mtl_handle shared_st_handle;
+extern unsigned int active_session_cnt;
+
+static int tx_st20p_frame_available(void *priv) {
+  KahawaiMuxerContext *s = priv;
+
+  pthread_mutex_lock(&(s->get_frame_mutex));
+  pthread_cond_signal(&(s->get_frame_cond));
+  pthread_mutex_unlock(&(s->get_frame_mutex));
+
+  return 0;
+}
+
+static int tx_st20p_frame_done(void* priv, struct st_frame *frame)
+{
+    KahawaiMuxerContext* s = priv;
+
+    if (ST_FRAME_STATUS_COMPLETE == frame->status )
+    {
+        frame_tx_completed++;
+        /* s->frame_counter is the number of frames sent to MTL, 
+          -1 here because some times we cannot get all notifications from MTL
+        */
+        if (frame_tx_completed == (s->frame_counter - 1))
+        {
+            frame_tx_completed = 0;
+            s->frame_counter = 0;
+            tx_completed = true;
+        }
+    }
+
+    return 0;
+}
+
+static int kahawai_write_header(AVFormatContext *ctx) {
+  KahawaiMuxerContext *s = ctx->priv_data;
+  AVStream *st = NULL;
+
+  int packet_size = 0;
+
+  int ret = 0;
+
+  struct mtl_init_params param;
+  struct st20p_tx_ops ops_tx;
+
+  av_log(ctx, AV_LOG_VERBOSE, "kahawai_write_header triggered\n");
+
+  memset(&param, 0, sizeof(param));
+  memset(&ops_tx, 0, sizeof(ops_tx));
+
+  if ((NULL == s->port) || (strlen(s->port) > MTL_PORT_MAX_LEN)) {
+    av_log(ctx, AV_LOG_ERROR, "Invalid port info\n");
+    return AVERROR(EINVAL);
+  }
+  param.num_ports = 1;
+  strncpy(param.port[MTL_PORT_P], s->port, MTL_PORT_MAX_LEN);
+  ops_tx.port.num_port = 1;
+  strncpy(ops_tx.port.port[MTL_PORT_P], s->port, MTL_PORT_MAX_LEN);
+
+  if (NULL == s->local_addr) {
+    av_log(ctx, AV_LOG_ERROR, "Invalid local IP address\n");
+    return AVERROR(EINVAL);
+  } else if (sscanf(s->local_addr, "%hhu.%hhu.%hhu.%hhu",
+                    &param.sip_addr[MTL_PORT_P][0],
+                    &param.sip_addr[MTL_PORT_P][1],
+                    &param.sip_addr[MTL_PORT_P][2],
+                    &param.sip_addr[MTL_PORT_P][3]) != MTL_IP_ADDR_LEN) {
+    av_log(ctx, AV_LOG_ERROR, "Failed to parse local IP address: %s\n",
+           s->local_addr);
+    return AVERROR(EINVAL);
+  }
+
+  param.tx_sessions_cnt_max = s->session_cnt;
+  param.rx_sessions_cnt_max = 0;
+  param.flags = MTL_FLAG_BIND_NUMA | MTL_FLAG_DEV_AUTO_START_STOP; // FIXME
+  param.log_level = MTL_LOG_LEVEL_DEBUG; // log level. ERROR, INFO, WARNING
```

Comment:
I think the default log level shall be INFO.

Reply:
Changed log level to INFO